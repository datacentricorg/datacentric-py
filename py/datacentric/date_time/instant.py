# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from abc import ABC, abstractmethod
from typing import Union, Optional, Tuple
import datetime as dt
import pytz


class Instant(dt.datetime, ABC):
    """
    Represents an instant on the global timeline in UTC timezone, with
    one millisecond resolution, and provides conversion to and from:

    * Int year, month, day, hour, minute, second, and millisecond in UTC
    * Int value of milliseconds since the Unix epoch
    * ISO string in yyy-mm-ddThh:mm:ss.fffZ format
    * Python dt.datetime
    * Pandas pd.timestamp
    """

    __unix_epoch: dt.datetime = dt.datetime.fromtimestamp(0, pytz.UTC)

    # --- METHODS

    @abstractmethod
    def abstract_class_guard(self) -> None:
        """
        Guard method to prevent this abstract base class from
        being instantiated.
        """
        pass

    # --- CLASS

    @classmethod
    def from_fields(cls,
                    year: int,
                    month: int,
                    day: int,
                    hour: int,
                    minute: int,
                    second: int,
                    millisecond: Optional[int] = None) -> Union[dt.datetime, 'Instant']:
        """
        Creates Instant represented as dt.datetime to millisecond precision from
        the the fields from year to millisecond.

        Millisecond field is optional. Zero value will be assumed if not specified.
        """

        # If millisecond is not specified, assume 0
        if millisecond is None:
            millisecond = 0

        # Create dt.datetime in UTC timezone
        result: dt.datetime = dt.datetime(
            year,
            month,
            day,
            hour,
            minute,
            second,
            1000 *
            millisecond,
            pytz.UTC)
        return result

    @classmethod
    def to_fields(cls, value: Union[dt.datetime, 'Instant']) -> Tuple[int, int, int, int, int, int, int]:
        """
        Convert Instant to millisecond precision
        the tuple (hour, minute, second, millisecond).
        """

        # Perform fast validation to detect values out of range.
        # This will not detect errors such as Feb 31
        cls.validate(value)

        # Calculate millisecond from microsecond
        millisecond: int = round(value.microsecond / 1000.0)

        # Return the tuple
        return value.year, value.month, value.day, value.hour, value.minute, value.second, millisecond

    @classmethod
    def from_str(cls, value: str) -> Union[dt.datetime, 'Instant']:
        """
        Convert from string in ISO format with UTC (Z) timezone suffix
        and up to 7 digits after decimal points for seconds.

        This method rounds the result to whole milliseconds.

        Examples:

        2003-05-01T10:15:30Z
        2003-05-01T10:15:30.1Z
        ...
        2003-05-01T10:15:30.1234567Z (will be rounded to whole milliseconds)
        """

        if not value.endswith('Z'):
            raise Exception(f'String {value} passed to Instant.from_str(...) method '
                            f'must end with capital Z that indicates UTC timezone.')

        # Convert to datetime and set UTC timezone
        dtime_from_str: dt.datetime
        if '.' in value:
            # Has milliseconds
            dtime_from_str = dt.datetime.strptime(
                value, '%Y-%m-%dT%H:%M:%S.%fZ')
        else:
            # Does not have milliseconds
            dtime_from_str = dt.datetime.strptime(value, '%Y-%m-%dT%H:%M:%SZ')

        # Set tzinfo to the standard UTC singleton from pytz module
        result: dt.datetime = dtime_from_str.replace(tzinfo=pytz.UTC)

        # Round the result to whole milliseconds
        result = cls.round(result)
        return result

    @classmethod
    def to_str(cls, value: Union[dt.datetime, 'Instant']) -> str:
        """
        Convert to string in ISO format with UTC (Z) timezone suffix
        and 3 digits after decimal points for seconds, irrespective of
        how many digits are actually required.

        Example:

        2003-05-01T10:15:30.500Z
        """

        # Ensure the argument is in UTC and does not have fractional
        # milliseconds
        cls.validate(value)

        # Convert to string in ISO format with UTC (Z) timezone suffix
        # and 3 digits after decimal points for seconds, irrespective of
        # how many digits are actually required.
        result_to_microseconds: str = value.strftime('%Y-%m-%dT%H:%M:%S.%f')
        result: str = result_to_microseconds[:-3] + 'Z'
        return result

    @classmethod
    def from_unix_millis(
            cls, unix_millis: int) -> Union[dt.datetime, 'Instant']:
        """
        Convert milliseconds since Unix epoch Instant stored as dt.datetime
        in UTC timezone.
        """
        # Convert milliseconds from Unix epoch to seconds, then construct the date
        # and replace timezone with pytz.UTC which is more standard
        return dt.datetime.utcfromtimestamp(
            unix_millis / 1000.0).replace(tzinfo=pytz.UTC)

    @classmethod
    def to_unix_millis(cls, value: Union[dt.datetime, 'Instant']) -> int:
        """
        Convert Instant stored as dt.datetime in UTC timezone to
        milliseconds since Unix epoch, rounding the value to
        whole milliseconds.

        Error message if the argument has fractional milliseconds,
        use Instant.round(value) to remove.
        """

        # Ensure the argument is in UTC and does not have fractional
        # milliseconds
        cls.validate(value)

        # Convert to milliseconds since Unix epoch and round to whole
        # milliseconds. Because we cannot subtract datetimes with two
        # different instances of tzinfo, even if both are UTC, we
        # cannot use a class variable for unix_epoch. It has to be
        # created in the argument timezone instead.
        unix_epoch: dt.datetime = dt.datetime.fromtimestamp(0, value.tzinfo)
        result: int = round((value - unix_epoch).total_seconds() * 1000)
        return result

    @classmethod
    def round(cls, value: Union[dt.datetime, 'Instant']
              ) -> Union[dt.datetime, 'Instant']:
        """
        Round the argument to whole milliseconds.

        This method checks that the argument is in UTC.
        """

        # Check that Instant is in UTC before rounding
        cls.check_in_utc(value)
        rounded_microsecond: int = round(value.microsecond / 1000.0) * 1000

        result: dt.datetime = dt.datetime(
            value.year,
            value.month,
            value.day,
            value.hour,
            value.minute,
            value.second,
            rounded_microsecond,
            value.tzinfo
        )
        return result

    @classmethod
    def validate(cls, value: Union[dt.datetime, 'Instant']) -> None:
        """
        Raise exception if one of the following is true:

        * Instant is not None and has timezone other than UTC
        * Instant is not None and has no timezone
        * Instant is not None and has fractional milliseconds
          (use Instant.round(...) to remove)
        """

        # Check that either value is None, or timezone is UTC
        cls.check_in_utc(value)

        # Check that either value is None, or Instant has no fractional milliseconds
        cls.check_whole_millis(value)

    @classmethod
    def check_in_utc(cls, value: Union[dt.datetime, 'Instant']) -> None:
        """
        Raise exception if value is not None, and is not in UTC timezone.

        This method does not check if the argument has fractional Unix
        milliseconds.
        """

        # If None, return before doing other checks
        if value is None:
            return

        # Check that tzinfo is not None
        if value.tzinfo is None:
            raise Exception(f'Instant {cls.__to_str_lenient(value)} has tzinfo=None. '
                            f'Instant values must have tzinfo that has zero offset '
                            f'to UTC, such as pytz.UTC')

        # Because:
        #
        # * Time zone info (tzinfo) is abstract class in Python
        # * There are several different singletons representing UTC, and
        # * Comparison will will fail if different instances are passed,
        #
        # we must use UTC offset instead of comparing timezone value.
        if value.utcoffset().microseconds != 0:
            raise Exception(f'Instant {cls.__to_str_lenient(value)} is not in UTC. '
                            f'Instant values must have tzinfo that has zero offset '
                            f'to UTC, such as pytz.UTC')

    @classmethod
    def check_whole_millis(cls, value: Union[dt.datetime, 'Instant']) -> None:
        """
        Return true if value is either None, or has no fractional milliseconds.

        This method does not check if the argument is in UTC.
        """

        # If None, return before doing other checks
        if value is None:
            return

        # Convert to milliseconds since Unix epoch and round to whole
        # milliseconds
        unix_epoch: dt.datetime = dt.datetime.fromtimestamp(0, value.tzinfo)
        unix_millis: float = (value - unix_epoch).total_seconds() * 1000
        rounded_unix_millis: int = round(unix_millis)

        # Return false if more than roundoff tolerance
        if abs(unix_millis - rounded_unix_millis) >= 1e-10:
            raise Exception(f'Instant {cls.__to_str_lenient(value)} has fractional milliseconds. Valid '
                f'Instant values must have whole milliseconds to ensure that no loss of precision '
                f'occurs during serialization roundtrip. Use Instant.round(...) to remove.')

    @classmethod
    def __to_str_lenient(cls, value: Union[dt.datetime, 'Instant']) -> str:
        """
        This method performs no validation before converting to string.
        It should be used for error messages only.
        """

        # Lenient conversion to string without controlling
        # for timezone or precision. Use for error messages
        # only.
        result: str = value.strftime('%Y-%m-%dT%H:%M:%S.%f')
        return result
