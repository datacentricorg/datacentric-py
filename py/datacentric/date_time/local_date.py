# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from abc import ABC, abstractmethod
from typing import Union
import datetime as dt


class LocalDate(int, ABC):
    """
    Represents a date within the calendar, with no reference to a particular
    time zone or time of day, and provides conversion to and from:

    * Integer year, month, and day of month
    * ISO string in yyyy-mm-dd format
    * Python dt.date
    """

    @staticmethod
    def to_iso_str(iso_int: int) -> str:
        """Convert to string in ISO yyyy-mm-dd format."""
        # We could directly format the output to string, however
        # this step will check that it is a valid date
        d: dt.date = LocalDate.to_date(iso_int)
        result: str = d.isoformat()
        return result

    @staticmethod
    def to_date(iso_int: int) -> dt.date:
        """Convert to dt.date."""
        value: int = iso_int
        year: int = value // 10_000
        value -= year * 10_000
        month: int = value // 100
        value -= month * 100
        day: int = value
        result: dt.date = dt.date(year, month, day)
        return result

    @staticmethod
    def from_date(date: dt.date) -> Union[int, 'LocalDate']:
        """Convert to int in ISO yyyymmdd format."""
        return 10_000 * date.year + 100 * date.month + date.day

    @staticmethod
    def from_iso_str(iso_string: str) -> Union[int, 'LocalDate']:
        """Convert from str in yyyy-mm-dd format to int in ISO yyyymmdd format."""
        if iso_string.endswith('Z'):
            raise Exception(f'String {iso_string} passed to LocalDate ctor must not end with capital Z that '
                            f'indicates UTC timezone because LocalDate does not include timezone.')

        # Convert from string in yyyy-mm-dd format
        date_from_str: dt.date = dt.date.fromisoformat(iso_string)

        # Convert to int in ISO yyyymmdd format
        return 10_000 * date_from_str.year + 100 * date_from_str.month + date_from_str.day

    @staticmethod
    def from_ints(year: int, month: int, day: int) -> Union[int, 'LocalDate']:
        """Convert int year, month and day to readable int format."""

        # TODO: improve check
        if not (1970 <= year <= 9999):
            raise Exception(f'year should be in 1970..9999 range.')
        if not (1 <= month <= 12):
            raise Exception(f'month should be in 1..12 range.')
        if not (1 <= month <= 31):
            raise Exception(f'day should be in 1..31 range.')

        # Convert to int in ISO yyyymmdd format
        return 10_000 * year + 100 * month + day

    @abstractmethod
    def abstract_class_guard(self) -> None:
        """
        Guard method to prevent this abstract base class from
        being instantiated.
        """
        pass
