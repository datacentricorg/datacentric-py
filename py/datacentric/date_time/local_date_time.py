# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from abc import ABC, abstractmethod
from typing import Union, Optional, Tuple
import datetime as dt


class LocalDateTime(int, ABC):
    """
    Represents datetime to millisecond precision, with no reference to a
    particular time zone, and provides conversion to and from:

    * Integer year, month, day, hour, minute, second, and optional millisecond
    * Integer to millisecond precision in yyyymmddhhmmssfff format
    * ISO string to millisecond precision in yyyy-mm-ddThh:mm:ss.fff format;
      seconds always have 3 digits after the decimal point, irrespective
      of how many digits are actually required.
    * Python dt.time without timezone
    """

    # --- METHODS

    @abstractmethod
    def abstract_class_guard(self) -> None:
        """
        Guard method to prevent this abstract base class from
        being instantiated.
        """
        pass

    # --- STATIC

    @classmethod
    def from_fields(cls, year: int, month: int, day: int, hour: int = 0,
                    minute: int = 0, second: int = 0, millisecond: Optional[int] = None) -> Union[int, 'LocalDateTime']:
        """
        Convert hour to millisecond fields to LocalDateTime represented in yyyymmddhhmmssfff format.

        Millisecond field is optional. Zero value will be assumed if not specified.
        """

        # If millisecond is not specified, assume 0
        if millisecond is None:
            millisecond = 0

        # Convert to LocalDateTime represented in yyyymmddhhmmssfff format
        result: int = 1000_00_00_00_00_00 * year + 1000_00_00_00_00 * month + 1000_00_00_00 * day + \
                      1000_00_00 * hour + 1000_00 * minute + 1000 * second + millisecond

        # Validate and return
        cls.validate(result)
        return result

    @classmethod
    def validate(cls, value: Union[int, 'LocalDateTime']) -> None:
        """
        Raise exception if the argument is not None, and is not an
        int in yyyymmddhhmmssfff format.

        This fast validation method will not detect errors such as Feb 31.
        """

        # If None, return before doing other checks
        if value is None:
            return

        # Convert to tuple
        year, month, day, hour, minute, second, millisecond = cls.__to_fields_lenient(value)

        if not (1970 <= year <= 9999):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The year {year} should be in 1970 to 9999 range.')
        if not (1 <= month <= 12):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The month {month} should be in 1 to 12 range.')
        if not (1 <= day <= 31):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The day {day} should be in 1 to 31 range.')
        if not (0 <= hour <= 23):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The hour {hour} should be in 0 to 23 range.')
        if not (0 <= minute <= 59):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The minute {minute} should be in 0 to 59 range.')
        if not (0 <= second <= 59):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The second {second} should be in 0 to 59 range.')
        if not (0 <= millisecond <= 999):
            raise Exception(f'LocalDateTime {value} is not in yyyymmddhhmmssfff format. '
                            f'The millisecond {millisecond} should be in 0 to 999 range.')

    @classmethod
    def __to_fields_lenient(cls, value: Union[int, 'LocalDateTime']) -> Tuple[int, int, int, int, int, int, int]:
        """
        Convert LocalDateTime represented as int in yyyymmddhhmmssfff format to
        the tuple (year, month, day, hour, minute, second, millisecond).

        This method does not perform validation of its argument.
        """

        year: int = value // 1000_00_00_00_00_00
        value -= year * 1000_00_00_00_00_00
        month: int = value // 1000_00_00_00_00
        value -= month * 1000_00_00_00_00
        day: int = value // 1000_00_00_00
        value -= day * 1000_00_00_00
        hour: int = value // 1000_00_00
        value -= hour * 1000_00_00
        minute: int = value // 1000_00
        value -= minute * 1000_00
        second: int = value // 1000
        value -= second * 1000
        millisecond: int = value

        return year, month, day, hour, minute, second, millisecond
