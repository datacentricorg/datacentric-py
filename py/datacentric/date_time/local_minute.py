# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations
import datetime as dt

# TODO - implementation is incomplete


class LocalMinute:
    """
    LocalMinute is an immutable struct representing a time of day
    to one minute precision, with no reference to a particular calendar,
    time zone or date.

    This class is not part of NodaTime but is inspired by NodaTime.LocalTime
    and follows the NodaTime naming conventions.
    """

    hour: int
    minute: int

    def __init__(self, hour: int, minute: int):
        """
        Creates local time to one minute precision from the specified
        hour and minute.
        """

        if hour < 0 or hour > 23:
            raise AttributeError(f'Hour {hour} is not between 0 and 23.')
        if minute < 0 or minute > 59:
            raise AttributeError(f'Minute {minute} is not between 0 and 59.')
        self.hour = hour
        self.minute = minute

    def to_time(self) -> dt.time:
        return dt.time(self.hour, self.minute)

    @property
    def minute_of_day(self) -> int:
        return self.hour * 60 + self.minute

    def __eq__(self, other):
        """
        Compares two local times for equality, by checking whether
        they represent the exact same local time, down to the tick.
        """
        if isinstance(other, LocalMinute):
            return self.minute_of_day == other.minute_of_day
        return NotImplemented

    def __lt__(self, other: LocalMinute):
        """
        Compares two LocalMinute values to see if the left one
        is strictly earlier than the right one.
        """
        return self.minute_of_day < other.minute_of_day

    def __le__(self, other: LocalMinute):
        """
        Compares two LocalMinute values to see if the left one
        is earlier than or equal to the right one.
        """
        return self.minute_of_day <= other.minute_of_day

    def __gt__(self, other: LocalMinute):
        """
        Compares two LocalMinute values to see if the left one
        is strictly later than the right one.
        """
        return self.minute_of_day > other.minute_of_day

    def __ge__(self, other: LocalMinute):
        """
        Compares two LocalMinute values to see if the left one
        is later than or equal to the right one.
        """
        return self.minute_of_day >= other.minute_of_day

    def __str__(self):
        """
        Convert LocalMinute to ISO 8601 string in hh:mm format.
        """
        return f'{self.hour:02}:{self.minute:02}'
