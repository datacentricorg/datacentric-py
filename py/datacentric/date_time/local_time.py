# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations
from typing import Union, Optional
import locale
import datetime as dt


class LocalTime:
    """
    Represents time of day to millisecond precision, with no reference to a
    particular date or time zone, and provides conversion to and from:

    * Integer hour, minute, second, and optional millisecond
    * Integer to millisecond precision in ISO hhmmssfff format
    * ISO string to millisecond precision in hh:mm:ss.fff format
    * Python dt.time
    """

    __slots__ = ('__iso_int',)

    __iso_int: int

    def __init__(self,
                 hour_or_value: Union[int, str, dt.time],
                 minute: Optional[int] = None,
                 second: Optional[int] = None,
                 millisecond: Optional[int] = None):
        """
        Creates LocalTime from the specified arguments. The options for the
        arguments include:

        * Integer hour, minute, second, and optional millisecond
        * Integer to millisecond precision in ISO hhmmssfff format
        * ISO string to millisecond precision in hh:mm:ss.fff format
        * Python dt.time
        """

        __iso_int = None
        """Local time to millisecond precision as int in ISO hhmmssfff format."""

        # Becuase Python has no method overloads, we have to use default ctor
        # arguments and determine the type dynamically. This flag is true
        # if first argument is used to pass the entire value, in which case
        # all other arguments (month, etc.) must be None
        remaining_args_must_be_none: bool = True

        if hour_or_value is None:
            raise Exception('In LocalTime constructor, first argument must not be None.')

        # Determine what is passed as first ctor argument
        if isinstance(hour_or_value, int):

            # Either hour, or the entite time in ISO hhmmssfff format
            hour_or_iso_int: int = hour_or_value

            if (minute is None) and (second is None) and (millisecond is None):

                # The sole int argument must be time in ISO hhmmssfff format
                if not (0 <= hour_or_iso_int <= 235959999):
                    raise Exception(f'The sole argument of LocalTime constructor {hour_or_iso_int} is not '
                                    f'integer in ISO hhmmssfff format.')
                self.__iso_int = hour_or_iso_int

            else:

                # If not the only argument, first argument is hour
                hour: int = hour_or_iso_int

                if not (0 <= hour_or_iso_int <= 23):
                    raise Exception(f'In LocalTime constructor, minute is specified but first '
                                    f'argument {hour} is not an integer in the range from 0 to 23.')
                if minute is None:
                    raise Exception(f'In LocalTime constructor, minute is None.')
                if not (0 <= minute <= 59):
                    raise Exception(f'In LocalTime constructor, minute {minute} is specified but is '
                                    f'not an integer in the range from 0 to 59.')
                if second is None:
                    raise Exception(f'In LocalTime constructor, second is None.')
                if not (0 <= second <= 59):
                    raise Exception(f'In LocalTime constructor, second {second} is specified but is '
                                    f'not an integer in the range from 0 to 59.')

                # This is one type of input where remaning args are used
                remaining_args_must_be_none = False

                if millisecond is None:
                    # If millisecond is not specified, assume 0
                    millisecond = 0
                else:
                    # Otherwise check the range
                    if not (0 <= millisecond <= 9999):
                        raise Exception(f'In LocalTime constructor, millisecond is specified but is '
                                        f'not an integer in the range from 0 to 999.')

                # Convert to int in ISO hhmmssfff format
                self.__iso_int = 10_000_000 * hour + 100_000 * minute + 1000 * second + millisecond

        elif isinstance(hour_or_value, str):

            # If argument is a string, it must use ISO hh:mm:ss.fff format
            iso_string: str = hour_or_value
            if iso_string.endswith('Z'):
                raise Exception(f'String {iso_string} passed to LocalTime ctor must not end with capital Z that '
                                f'indicates UTC timezone because LocalTime does not include timezone.')

            time_whole_seconds: dt.time
            millisecond_int: int
            if '.' in iso_string:

                # Has milliseconds, onvert from string in hh:mm:ss.fff format
                # by converting whole seconds first, then adding milliseconds
                iso_string_tokens = iso_string.split('.', 2)
                iso_string_whole_seconds: str = iso_string_tokens[0]
                time_whole_seconds = dt.time.fromisoformat(iso_string_whole_seconds)

                # Validate fractional seconds
                fractional_second_str: str = '0.' + iso_string_tokens[1]
                millisecond_int = round(1000.0 * locale.atof(fractional_second_str))
                if not (0 <= millisecond_int <= 9999):
                    raise Exception(f'In LocalTime constructor from string, fractional second {fractional_second_str} '
                                    f'is specified but is not in the range from 0 (inclusive) to 1 (exclusive).')

            else:

                # Does not have milliseconds
                time_whole_seconds = dt.time.fromisoformat(iso_string)
                millisecond_int = 0

            # Convert to int in ISO hhmmssfff format
            self.__iso_int = 10_000_000 * time_whole_seconds.hour + 100_000 * time_whole_seconds.minute \
                             + 1000 * time_whole_seconds.second + millisecond_int

        elif isinstance(hour_or_value, dt.time):

            # First argument is dt.datetime or pd.Timestamp
            time_arg: dt.time = hour_or_value

            # Convert to int in ISO hhmmssfff format
            self.__iso_int = 10_000_000 * time_arg.hour + 100_000 * time_arg.minute \
                             + 1000 * time_arg.second + round(time_arg.microsecond / 1000.0)

        else:
            raise Exception(f'First argument of LocalTime constructor {hour_or_value} must be one of '
                            f'(a) hour, (b) integer in ISO hhmmssfff format, (c) string in ISO hh:mm:ss.fff format, '
                            f'or (d) dt.time.')

        # Depending on what is passed as the first argument, validate the remaining arguments
        if remaining_args_must_be_none:

            # If the first argument is not year, the remaining arguments must be None
            if minute is not None:
                raise Exception(f'In LocalTime constructor, first argument {hour_or_value} is the entire date '
                                f'rather than year, in which case minute={minute} must be None.')
            if second is not None:
                raise Exception(f'In LocalTime constructor, first argument {hour_or_value} is the entire date '
                                f'rather than year, in which case second={second} must be None.')
            if millisecond is not None:
                raise Exception(f'In LocalTime constructor, first argument {hour_or_value} is the entire date '
                                f'rather than year, in which case millisecond={millisecond} must be None.')

    def __str__(self) -> str:
        """Convert to string in ISO hh:mm:ss.fff format."""
        # We could directly format the output to string, however
        # this step will check that it is a valid time
        t: dt.time = self.to_time()
        result: str = t.isoformat()

        """
        Convert to string in ISO format to millisecond precision, for example:

        10:15:30.5

        Fractional seconds are omitted if zero.
        """
        value: int = self.__iso_int
        hour: int = value // 10_000_000
        value -= hour * 10_000_000
        minute: int = value // 100_000
        value -= minute * 100_000
        second: int = value // 1_000
        value -= second * 1_000
        millisecond: int = value

        # Convert whole seconds separately from milliseconds
        time_whole_seconds: dt.time = dt.time(hour, minute, second)
        time_whole_seconds_str: str = time_whole_seconds.isoformat()
        if millisecond == 0:
            # No milliseconds
            return time_whole_seconds_str
        else:
            # Include milliseconds without the trailing zeroes
            millis_str: str = str(millisecond / 1000.0).lstrip('0.')
            local_time_str: str = time_whole_seconds_str + '.' + millis_str
            return local_time_str

    def to_iso_int(self) -> int:
        """Convert to int in ISO hhmmssfff format."""
        return self.__iso_int

    def to_time(self) -> dt.time:
        """Convert to dt.time."""
        value: int = self.__iso_int
        hour: int = value // 10_000_000
        value -= hour * 10_000_000
        minute: int = value // 100_000
        value -= minute * 100_000
        second: int = value // 1_000
        value -= second * 1_000
        millisecond: int = value
        result: dt.time = dt.time(hour, minute, second, 1000 * millisecond)
        return result

    def __eq__(self, other):
        """
        True if lhs and rhs represent the same moment in time.

        Returns NotImplemented if rhs is None or not a LocalTime.
        """
        if isinstance(other, LocalTime):
            return self.__iso_int == other.__iso_int
        return NotImplemented

    def __lt__(self, other: LocalTime):
        """
        True if lhs is strictly earlier than rhs.

        Error message if rhs is None.
        """
        return self.__iso_int < other.__iso_int

    def __le__(self, other: LocalTime):
        """
        True if lhs is earlier than or equal to rhs.

        Error message if rhs is None.
        """
        return self.__iso_int <= other.__iso_int

    def __gt__(self, other: LocalTime):
        """
        True if lhs is strictly later than rhs.

        Error message if rhs is None.
        """
        return self.__iso_int > other.__iso_int

    def __ge__(self, other: LocalTime):
        """
        True if lhs is later than or equal to rhs.

        Error message if rhs is None.
        """
        return self.__iso_int >= other.__iso_int
