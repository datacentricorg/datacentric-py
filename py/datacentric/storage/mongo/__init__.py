from datacentric.storage.mongo.mongo_data_source import MongoDataSource
from datacentric.storage.mongo.temporal_mongo_data_source import TemporalMongoDataSource
from datacentric.storage.mongo.temporal_mongo_query import TemporalMongoQuery