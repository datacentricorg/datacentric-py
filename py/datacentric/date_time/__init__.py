from datacentric.date_time.iso_day_of_week import IsoDayOfWeek
from datacentric.date_time.local_minute import LocalMinute
