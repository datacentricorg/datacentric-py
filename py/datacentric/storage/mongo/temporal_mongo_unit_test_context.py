# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from datacentric.storage.temporal_id import empty_id
from datacentric.storage.unit_test_context import UnitTestContext
from datacentric.storage.mongo.temporal_mongo_data_source import TemporalMongoDataSource
from datacentric.storage.env_type import EnvType
from datacentric.storage.versioning_method import VersioningMethod


class TemporalMongoUnitTestContext(UnitTestContext):
    """
    TemporalMongoUnitTestContext is the context for use in test fixtures
    that require a temporal Mongo data source.

    It extends UnitTestContext by creating an empty test
    database specific to the test method, and deleting
    it after the test exits. The context creates Common
    dataset in the database and assigns its TemporalId to
    the DataSet property of the context.

    If the test sets KeepTestData = true, the data is retained
    after the text exits. This data will be cleared on the
    next launch of the test.

    For tests that do not require a data source, use UnitTestContext.
    """

    __slots__ = ()

    def __init__(self):
        """Inspect call stack to set properties."""
        super().__init__()

        # Create and initialize data source with TEST environment type.
        #
        # This does not create the database until the data source
        # is actually used to access data.

        # Create data source specified as generic argument
        self.data_source = TemporalMongoDataSource(
            env_type=EnvType.Test,
            env_group=self.test_module_name,
            env_name=self.test_method_name,
            versioning_method=VersioningMethod.Temporal
        )

        # Delete (drop) the database to clear the existing data
        self.data_source.delete_db()

        # Create Common dataset and assign it to data_set property of this context
        self.data_set = empty_id
