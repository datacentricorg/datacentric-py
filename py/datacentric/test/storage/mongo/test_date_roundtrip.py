# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest
import datetime as dt
from datacentric.date_time.local_date import LocalDate
from datacentric.date_time.local_time import LocalTime
from datacentric.date_time.local_minute import LocalMinute
from datacentric.date_time.local_date_time import LocalDateTime
from datacentric.date_time.instant import Instant
from datacentric.test.storage.date_sample import DatesSample
from datacentric.storage.mongo.temporal_mongo_unit_test_context import TemporalMongoUnitTestContext


class TestDateRoundTrip(unittest.TestCase):
    """Tests for TestDateRoundTrip."""

    def test_smoke(self):
        """Smoke test to check equality after data source save/load."""

        with TemporalMongoUnitTestContext() as context:
            data_set0 = context.data_source.create_data_set('DataSet0')

            rec = DatesSample()
            rec.record_name = 'Sample'
            rec.local_date_element = LocalDate.from_fields(2003, 5, 1)
            rec.local_time_element = LocalTime.from_fields(10, 15, 30)  # 10:15:30
            rec.local_minute_element = LocalMinute.from_fields(10, 15)  # 10:15
            rec.local_date_time_element = LocalDateTime.from_fields(2003, 5, 1, 10, 15)  # 2003-05-01T10:15:00
            rec.instant_element = Instant.from_fields(2003, 5, 1, 10, 15, 0)

            rec.date_element = dt.date(2003, 5, 1)
            rec.time_element = dt.time(10, 15, 30)
            rec.date_time_element = dt.datetime(2003, 5, 1, 10, 15)

            context.data_source.save_one(DatesSample, rec, data_set0)

            loaded = context.data_source.load_by_key(DatesSample, rec.to_key(), data_set0)

            self.assertEqual(rec.record_name, loaded.record_name)
            self.assertEqual(rec.local_minute_element, loaded.local_minute_element)
            self.assertEqual(rec.local_date_element, loaded.local_date_element)
            self.assertEqual(rec.local_time_element, loaded.local_time_element)
            self.assertEqual(rec.local_date_time_element, loaded.local_date_time_element)
            self.assertEqual(rec.instant_element, loaded.instant_element)

            self.assertEqual(rec.date_element, loaded.date_element)
            self.assertEqual(rec.time_element, loaded.time_element)
            self.assertEqual(rec.date_time_element, loaded.date_time_element)
