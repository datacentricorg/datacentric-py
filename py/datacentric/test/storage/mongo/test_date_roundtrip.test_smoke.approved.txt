Error: Exception Exception has occurred with message Cannot serialize type date.
